### PR TITLE
Include ARQ scheme #10

### LINKED ISSUE #10: Implement ARQ for data transmission

### LINKED ISSUE #42: Refactoring of mac_frag_get_fragment
Currently (5ffd0417ecce7a9df6af43c2844005a180fcd570), `mac_frag_get_fragment` has over 100 lines and deep intendation nesting.
Refactoring is desirable.

### PR DESCRIPTION
**Pull request by lukasostendorf**
This PR enhances the system with an additional MAC transmission mode.
Currently all MAC frames are sent in Unacknowledged mode, this PR allows to use a selective repeat ARQ in parallel to the existing Unacknowledged mode.

Fixes #10

### CODE REVIEW

File: src/mac/mac_messages.h [hunk 1]
```diff
@@ -40,18 +42,26 @@ typedef enum {
...
+  ul_data_ack,
   dl_data = 7,
   ul_req = 9,
   channel_quality,
   keepalive,
   control_ack,
   mcs_chance_req,
+  dl_data_ack,
   ul_data = 15
 } CtrlID_e;
```
Comment (dl1com) [inline src/mac/mac_messages.h:54]: Please assign enum values explicitly to avoid potential confusion due to different enummeration when using different compilers.

File: src/mac/mac_bs.c [hunk 2]
```diff
@@ -692,6 +709,19 @@ void *mac_bs_tap_rx_th(void *arg) {
       MacDataFrame frame = dataframe_create(dev->bytes_rec);
       memcpy(frame->data, dev->buffer, dev->bytes_rec);
 
+      // Packet inspection: determine if this is TCP traffic
```
Comment (dl1com) [inline src/mac/mac_bs.c]: Also used in mac_ue.c -> move this functionality to a common place?

File: src/mac/mac_ue.c [hunk 3]
```diff
@@ -358,6 +374,19 @@ void *mac_ue_tap_rx_th(void *arg) {
     if (mac->tapdevice->bytes_rec > 0) {
       MacDataFrame frame = dataframe_create(mac->tapdevice->bytes_rec);
       memcpy(frame->data, mac->tapdevice->buffer, frame->size);
+      // Packet inspection: determine if this is TCP traffic
```
Comment (dl1com) [inline src/mac/mac_ue.c]: see comment in mac_bs.c

**Review by dl1com — COMMENTED** [review]
I guess it would be nice to have some more verbose documentation on a per-function level, especially the input and output parameters. I'd propose to use doxygen format or something similar.
No need to do everything at once, but all functions which where touched by this PR may be a good start :)

File: src/mac/mac_ue.c [hunk 4]
```diff
@@ -250,8 +250,11 @@ void mac_ue_run_scheduler(MacUE mac) {
             lchan_add_message(chan, msg);
             mac_msg_destroy(msg);
           }
-          MacMessage msg = mac_frag_get_fragment(mac->fragmenter,
-                                                 lchan_unused_bytes(chan), 1);
+          MacMessage msg = NULL;
+          if (mac_frag_has_fragment(mac->fragmenter)) {
```
Comment (dl1com) [inline src/mac/mac_ue.c]: As there is no else criteria here, it might also make sense to include the call of the "mac_frag_has_fragment()" check into mac_frag_get_fragment()?
Otherwise it will probably forgotten to be called before other (future) usages of ...get_fragment() as well.

File: src/mac/mac_messages.c [hunk 5]
```diff
@@ -409,8 +409,8 @@ void mac_msg_parse_mcs_change_req(MacMessage msg) {
 
 void mac_msg_parse_dl_data_ack(MacMessage msg) {
   msg->hdr.DLdataAck.ctrl_id = msg->type & 0b111;
-  msg->hdr.DLdataAck.ack_type = (msg->hdr_bin[0] >> 3) & 0b11;
-  msg->hdr.DLdataAck.seqNr = msg->hdr_bin[0] & 0b111;
+  msg->hdr.DLdataAck.ack_type = (msg->hdr_bin[0] >> 4) & 0b1;
+  msg->hdr.DLdataAck.seqNr = msg->hdr_bin[0] & 0b1111;
   msg->hdr.DLdataAck.fragNr = (msg->hdr_bin[1] >> 3) & 0b11111;
```
Comment (dl1com) [inline src/mac/mac_messages.c]: Is it possible to add a unit test for this?

File: src/mac/mac_messages.c [hunk 6]
```diff
@@ -154,8 +154,8 @@ MacMessage mac_msg_create_ul_data_ack(uint8_t ack_type, uint8_t seqNr,
   genericmsg->payload_len = 0;
 
   genericmsg->hdr_bin[0] = (ul_data_ack & 0b111) << 5;
-  genericmsg->hdr_bin[0] |= (ack_type & 0b11) << 3;
-  genericmsg->hdr_bin[0] |= (seqNr & 0b111);
+  genericmsg->hdr_bin[0] |= (ack_type & 0b1) << 4;
+  genericmsg->hdr_bin[0] |= (seqNr & 0b1111);
   genericmsg->hdr_bin[1] = (fragNr & 0b11111) << 3;
```
Comment (dl1com) [inline src/mac/mac_messages.c]: Is it possible to add a unit test for this?

File: src/mac/mac_messages.c [hunk 7]
```diff
@@ -262,8 +262,8 @@ MacMessage mac_msg_create_dl_data_ack(uint8_t ack_type, uint8_t seqNr,
   genericmsg->payload_len = 0;
 
   genericmsg->hdr_bin[0] = (dl_data_ack & 0b111) << 5;
-  genericmsg->hdr_bin[0] |= (ack_type & 0b11) << 3;
-  genericmsg->hdr_bin[0] |= (seqNr & 0b111);
+  genericmsg->hdr_bin[0] |= (ack_type & 0b1) << 4;
+  genericmsg->hdr_bin[0] |= (seqNr & 0b1111);
   genericmsg->hdr_bin[1] = (fragNr & 0b11111) << 3;
```
Comment (dl1com) [inline src/mac/mac_messages.c]: Is it possible to add a unit test for this?

File: src/mac/mac_messages.c [hunk 8]
```diff
@@ -366,8 +366,8 @@ void mac_msg_parse_timing_advance(MacMessage msg) {
 
 void mac_msg_parse_ul_data_ack(MacMessage msg) {
   msg->hdr.ULdataAck.ctrl_id = msg->type & 0b111;
-  msg->hdr.ULdataAck.ack_type = (msg->hdr_bin[0] >> 3) & 0b11;
-  msg->hdr.ULdataAck.seqNr = msg->hdr_bin[0] & 0b111;
+  msg->hdr.ULdataAck.ack_type = (msg->hdr_bin[0] >> 4) & 0b1;
+  msg->hdr.ULdataAck.seqNr = msg->hdr_bin[0] & 0b1111;
   msg->hdr.ULdataAck.fragNr = (msg->hdr_bin[1] >> 3) & 0b11111;
 }
```
Comment (dl1com) [inline src/mac/mac_messages.c]: Is it possible to add a unit test for this?

File: src/mac/mac_messages.c (hunk 6 above)
Comment (lukasostendorf) [reply src/mac/mac_messages.c]: There is a unit test for dl_data_ack and ul_data_ack messages included in this branch.

test_mac_msg_parse_ul_data_ack()
test_mac_msg_parse_dl_data_ack()

in test_mac_messages.c

But I think it may be better to use the boundary values seqNr=15 and fragNr=31 for the test instead of the currently randomly chosen seqNr=5 and fragNr=13

File: src/mac/mac_messages.h (hunk 1 above)
Comment (lukasostendorf) [reply src/mac/mac_messages.h:54]: Fixed in https://github.com/HAMNET-Access-Protocol/HNAP4PlutoSDR/pull/37/commits/5ffd0417ecce7a9df6af43c2844005a180fcd570

File: src/mac/mac_ue.c (hunk 4 above)
Comment (lukasostendorf) [reply src/mac/mac_ue.c]: This is a good idea, I've also included this in https://github.com/HAMNET-Access-Protocol/HNAP4PlutoSDR/pull/37/commits/5ffd0417ecce7a9df6af43c2844005a180fcd570

File: src/mac/mac_bs.c (hunk 2 above)
Comment (lukasostendorf) [reply src/mac/mac_bs.c]: This is now moved to mac_common.c

**Review by dl1com — COMMENTED** [review]
Nice work!
Just two more comments:

- Could you please add an unit test for `packet_inspect_is_tcpip`?
- Function `mac_frag_get_fragment` is of pretty high complexity (with over 100 lines) with a high level of indentations - I've opened an issue for refactoring: #42

**Review by dl1com — APPROVED** [review]
